core/sr-primitives/src/sr_arithmetic.rs

```diff
@@ -0,0 +1,1155 @@
+				);
+			}
+
+			#[test]
+			fn per_things_div_works() {
+				// normal
+				assert_eq!($name::from_percent(10) / $name::from_percent(20),
+					$name::from_percent(50)
+				);
+				assert_eq!($name::from_percent(10) / $name::from_percent(10),
+					$name::from_percent(100)
+				);
+				assert_eq!($name::from_percent(10) / $name::from_percent(0),
+					$name::from_percent(100)
+				);
+
+				// will not overflow
+				assert_eq!($name::from_percent(10) / $name::from_percent(5),
+					$name::from_percent(100)
+				);
+				assert_eq!($name::from_percent(100) / $name::from_percent(50),
+					$name::from_percent(100)
+				);
+			}
+		}
+	};
+}
+
+implement_per_thing!(
+	Percent,
+	test_per_cent,
+	[u32, u64, u128],
+	100u8,
+	u8,
+	u16,
+	"_Percent_",
+);
+implement_per_thing!(
+	Permill,
+	test_permill,
+	[u32, u64, u128],
+	1_000_000u32,
+	u32,
+	u64,
+	"_Parts per Million_",
+);
+implement_per_thing!(
+	Perbill,
+	test_perbill,
+	[u32, u64, u128],
+	1_000_000_000u32,
+	u32,
+	u64,
+	"_Parts per Billion_",
+);
+implement_per_thing!(
+	Perquintill,
+	test_perquintill,
+	[u64, u128],
+	1_000_000_000_000_000_000u64,
+	u64,
+	u128,
+	"_Parts per Quintillion_",
+);
+
+/// An unsigned fixed point number. Can hold any value in the range [-9_223_372_036, 9_223_372_036]
+/// with fixed point accuracy of one billion.
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(Encode, Decode, Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Fixed64(i64);
+
+/// The accuracy of the `Fixed64` type.
+const DIV: i64 = 1_000_000_000;
+
+impl Fixed64 {
+	/// creates self from a natural number.
+	///
+	/// Note that this might be lossy.
+	pub fn from_natural(int: i64) -> Self {
+		Self(int.saturating_mul(DIV))
+	}
+
+	/// Return the accuracy of the type. Given that this function returns the value `X`, it means
+	/// that an instance composed of `X` parts (`Fixed64::from_parts(X)`) is equal to `1`.
+	pub fn accuracy() -> i64 {
+		DIV
+	}
+
+	/// Raw constructor. Equal to `parts / 1_000_000_000`.
+	pub fn from_parts(parts: i64) -> Self {
+		Self(parts)
+	}
+
+	/// creates self from a rational number. Equal to `n/d`.
+	///
+	/// Note that this might be lossy.
+	pub fn from_rational(n: i64, d: u64) -> Self {
+		Self(
+			((n as i128).saturating_mul(DIV as i128) / (d as i128).max(1))
+			.try_into()
+			.unwrap_or(Bounded::max_value())
+		)
+	}
+
+	/// Performs a saturated multiply and accumulate by unsigned number.
+	///
+	/// Returns a saturated `int + (self * int)`.
+	pub fn saturated_multiply_accumulate<N>(&self, int: N) -> N
+	where
+		N: TryFrom<u64> + From<u32> + UniqueSaturatedInto<u32> + Bounded + Clone + Saturating +
+		ops::Rem<N, Output=N> + ops::Div<N, Output=N> + ops::Mul<N, Output=N> +
+		ops::Add<N, Output=N>,
+	{
+		let div = DIV as u64;
+		let positive = self.0 > 0;
+		// safe to convert as absolute value.
+		let parts = self.0.checked_abs().unwrap_or(Bounded::max_value()) as u64;
```

Comment:
if self.0 is i64::min_value then it will result in `9223372036854775807` instead of `9223372036854775808`
```suggestion
		let parts = self.0.checked_abs().map(|v| v as u64).unwrap_or(i64::max_value() as u64 + 1);
```

Reply:
true, only can happen in `i64::MIN`. Your scrutiny is impressive 👍